### ISSUE TITLE
tls-lwt cause a module clash with the Core library

### ISSUE THREAD

**Issue by avsm** [mirleft/ocaml-tls#376]
*impact:* cannot link any executables that use tls-lwt and Core libraries together

The Tls library uses pack to hide its internal modules, but Tls_lwt still scans the include directory of `lib` and spots the internal module cmis.

This means that `$ ocamlobjinfo _build/lwt/tls_lwt.cmi` imports a module called Core, which subsequently causes a global linking clash when trying to link Core.

The solution is to avoid scanning the internals of the `tls` library when building `tls-lwt`. I'm not sure how to do that in ocamlbuild with the existing _tags files, but resurrecting #367 would do the trick since jbuilder isolates libraries in a separate directory by default.

**Comment by hannesm** [comment]
I agree that this is annoying, but I've no time atm to fix it properly (see [david's comment](https://github.com/mirleft/ocaml-tls/pull/337#issuecomment-256101552)).  the proper fix includes to come up with an API for TLS (providing `tls.mli`, and prefix internal modules with `Tls_`).  since quite some libraries use `Tls.Core` these days, I'm cautious about just renaming it and thus breaking them now (esp. with another API breakage in the (hopefully soon) future).

**Comment by avsm** [comment]
It might be fixable with a myocamlbuild hack to stage the packed tls.mli into another directory, and then only include that directory in the _tags file.

**Comment by pveber** [comment]
Any workaround? I'd like to use `cohttp-lwt-unix` with `core` and was stopped by this issue.

**Comment by rgrinberg** [comment]
One workaround is to switch to dune ;) I know @hannesm is against it but perhaps we could publish a dual package into opam like `tls-core` that would conflict with the original tls. The only purpose of this package would be core compatibility.

**Comment by hannesm** [comment]
Dear @rgrinberg, "I know @hannesm is against it" is just wrong.  Switching to `dune` will not magically solve any problems.  You may want to read up on #337 and #367 again to re-discover what the actual issue is, briefly: missing an API for the core TLS library (and missing reasonable funding and manpower to do so).

**Comment by rgrinberg** [comment]
Dear @hannesm, the issue is as I understand it, is that tls-lwt makes references to an intermediate build artifact that it should not (`core.cmi`). We can quickly verify this:

```
$ ocamlobjinfo ~/.opam/4.05.0/lib/tls/tls_lwt.cmi | grep Core
        dca1769b4a50dcf5dbaac1279975de7a        Core
```

However, my port to dune fixes exactly this problem (without any magic!), and without breaking the API in any way

```
$ ocamlobjinfo _build/default/lwt/tls_lwt.cmi | grep Core
        9ec25decfa10e23cd95eb7cb06e46611        Tls__Core
```

In fact, if you use dune you will never run into a situation where your intermediate artifacts are visible where they shouldn't be.

You may want to read my PR and re-discover the issues that it solves. I'm sure that your users will appreciate this more than your condescension.

To briefly workaround this issue, until it's re-introduced yet again due to do the broken build system, you may consider applying the following patch:

```
diff --git a/lwt/tls_lwt.mli b/lwt/tls_lwt.mli
index a643172..8c7cff6 100644
--- a/lwt/tls_lwt.mli
+++ b/lwt/tls_lwt.mli
@@ -76,7 +76,7 @@ module Unix : sig
       be adjusted by [cert]. If [drop] is [true] (the default),
       application data received before the renegotiation finished is dropped. *)
   val reneg : ?authenticator:X509.Authenticator.a ->
-    ?acceptable_cas:X509.distinguished_name list -> ?cert:Config.own_cert ->
+    ?acceptable_cas:X509.distinguished_name list -> ?cert:Tls.Config.own_cert ->
     ?drop:bool -> t -> unit Lwt.t
 
   (** [epoch t] returns [epoch], which contains information of the
```

**Comment by hannesm** [comment]
@rgrinberg thanks for the investigative work, pushed, and released 0.9.1 as https://github.com/ocaml/opam-repository/pull/11481
